dotCMS/src/main/java/com/dotmarketing/startup/runonce/Task04315UpdateMultiTreePK.java

```diff
@@ -0,0 +1,127 @@
+import com.dotmarketing.db.DbConnectionFactory;
+import com.dotmarketing.exception.DotDataException;
+import com.dotmarketing.startup.AbstractJDBCStartupTask;
+
+import java.sql.Connection;
+import java.util.List;
+
+/**
+ * This task updates the definition of the {@code workflow_action} table, creates the new
+ * intermediate {@code workflow_action_step} table, and updates the definition of a table index:
+ * <ul>
+ * <li>A new column called {@code scheme_id} will be added to the {@code workflow_action} table.
+ * This allows actions to be assigned to Workflow Schemes and not to steps directly.</li>
+ * <li>The new {@code workflow_action_step} table allows an action to be associated to one or more
+ * steps (an N-N relationship).</li>
+ * <li>The index created for the {@code workflow_scheme_x_structure} column will be updated to NOT
+ * be unique.</li>
+ * </ul>
+ *
+ * @author Jose Castro
+ * @version 4.3.0
+ * @since Nov 1st, 2017
+ */
+public class Task04315UpdateMultiTreePK extends AbstractJDBCStartupTask{
+
+
+
+    final static String tableName ="multi_tree";
+    final static String indexName ="idx_multitree_index1";
+    final static String columnNames = "parent1, parent2, child, relation_type";
+
+    
+    
+    final static String updateRelationTypesSQL =   "update multi_tree set relation_type='" + MultiTree.LEGACY_RELATION_TYPE + "' where relation_type is null or relation_type=''";
+    
+    
+    final static String alterRelationTypePK =   "ALTER TABLE " + tableName + " ADD CONSTRAINT " + indexName + " PRIMARY KEY (" + columnNames + ")";
+    
+    final static String addIndexToMultiTree =   "CREATE INDEX  idx_multi_tree_page_relation on " + tableName + " (parent1, relation_type) ";
+    
+    @Override
+    public boolean forceRun() {
+        return Boolean.TRUE;
+    }
+
+    @Override
+    public void executeUpgrade() throws DotDataException {
+        
+        try {
+            Connection conn =DbConnectionFactory.getConnection();
+            
+            
+            // Drop PK
+            getPrimaryKey(conn,ImmutableList.of(tableName), true );
+            
+            // Drop Indexes
+            this.getIndexes(conn, ImmutableList.of(tableName), true );
+            
+            
+            DotConnect db = new DotConnect();
+            db.setSQL(updateRelationTypesSQL);
+            db.loadResult(conn);
+            
+            
+            db.setSQL(alterRelationTypePK);
+            db.loadResult(conn);
+            
+
+            
+            db.setSQL(addIndexToMultiTree);
+            db.loadResult(conn);
+            
+
+            
+
+
+        }
+        catch(Exception e) {
+            throw new DotDataException(e);
+        }
+
+    } // executeUpgrade.
+
+    @Override
+    public String getPostgresScript() {
+        // TODO Auto-generated method stub
+        return null;
+    }
+
+    @Override
+    public String getMySQLScript() {
+        // TODO Auto-generated method stub
```

Comment:
![INFO](https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO') Complete the task associated to this TODO comment. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1135)